lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -92,3 +105,377 @@ def gen_month_range(prefix, index, start, end):
+            Parameter list iterator
+        """
+        return iter(self.parameters)
+
+    def __str__(self):
+        return f"Schema<{str(self.parameters)}>"
+
+
+class ElasticBase(Resource):
+    """
+    ElasticBase A base class for Elasticsearch queries that allows subclasses
+    to provide customers pre- and post- processing.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger, schema: Schema):
+        """
+        __init__ Construct the base class
+
+        Args:
+            config: server configuration
+            logger: logger object
+            schema: API schema: for example,
+                    Schema(
+                        Parameter("user", ParamType.USER, required=True),
+                        Parameter("start", ParamType.DATE)
+                    )
+
+        NOTE: each class currently only supports one HTTP method, so the
+        required parameters doesn't distinguish. This could be expanded to
+        specify multiple schema by HTTP method if that becomes necessary.
+        """
+        self.logger = logger
+        self.config = config
+        self.prefix = self.config.get("Indexing", "index_prefix")
+        host = self.config.get("elasticsearch", "host")
+        port = self.config.get("elasticsearch", "port")
+        self.es_url = f"http://{host}:{port}"
+        self.schema = schema
+
+    def _get_user_term(self, user: str):
+        """Elasticsearch doesn't index (or search for) null values, and it's not
+        clear we really want to consider "owner": null to be "owned" datasets, in
+        any case. (Nor will they be possible beyond the transition period where we
+        may retain limited support for 0.69 data.)
+
+        Instead, let a query for data with a missing or null "user" parameter find
+        all data that's set to allow public access. This by default includes all
+        "unowned" data, but also data owned by a user that's been published.
+        """
+        if user:
+            term = {"authorization.owner": user}
+        else:
+            term = {"authorization.access": "public"}
+        return term
+
+    def _gen_month_range(
+        self, index: str, start: datetime, end: datetime
+    ) -> List[AnyStr]:
+        """
+        gen_month_range Construct a comma-separated list of index names
+        qualified by year and month suitable for use in the Elasticsearch
+        /_search query URI.
+        The month is incremented by 1 from "start" to "end"; for example,
+        gen_month_range('drb.', 'v4.run.', '2020-08', '2020-10') will result
+        in
+            'drb.v4.run.2020-08,drb.v4.run.2020-09,drb.v4.run.2020-10,'
+        This code is adapted from Javascript in the Pbench dashboard's
+        'moment_constants.js' getAllMonthsWithinRange function.
+        Args:
+            prefix: The Pbench server index prefix
+            index: The desired monthly index root
+            start: The start time
+            end: The end time
```

Comment:
I would like to believe that, between well-named parameters and type hints, nearly everything that I'm going to need as a reader is right there in the function signature.  For the few cases where I'll need more, it can be described in the header comment for the function.  Thus, there is no need for the header comment to describe the all the parameters every time; and, I think it's a savings if we don't.  (It allows us to fit more code on the screen, e.g.)